lib/SIL/TypeLowering.cpp

```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
   
   return boxType;
 }
+
+static void countNumberOfInnerFields(size_t &fieldsCount, SILModule &Module,
+                                     SILType Ty) {
+  if (auto *structDecl = Ty.getStructOrBoundGenericStruct()) {
+    if (!structDecl->hasFixedLayout(Module.getSwiftModule(),
+                                    ResilienceExpansion::Minimal)) {
+      return;
+    }
+    for (auto prop : structDecl->getStoredProperties()) {
+      SILType propTy = Ty.getFieldType(prop, Module);
+      size_t fieldsCountBefore = fieldsCount;
+      countNumberOfInnerFields(fieldsCount, Module, propTy);
+      if (fieldsCount == fieldsCountBefore) {
+        // size of Struct(BigStructType) == size of BigStructType()
+        // prevent counting its size as BigStructType()+1
+        ++fieldsCount;
+      }
+    }
+  } else if (Ty.is<TupleType>()) {
+    auto tupleTy = Ty.castTo<TupleType>();
+    for (auto elt : tupleTy->getElementTypes()) {
+      auto silElt = SILType::getPrimitiveObjectType(elt->getCanonicalType());
+      countNumberOfInnerFields(fieldsCount, Module, silElt);
+    }
+  } else if (auto *enumDecl = Ty.getEnumOrBoundGenericEnum()) {
+    if (enumDecl->isIndirect()) {
+      return;
+    }
+    if (!enumDecl->hasFixedLayout(Module.getSwiftModule(),
+                                  ResilienceExpansion::Minimal)) {
+      return;
+    }
+    size_t fieldsCountBefore = fieldsCount;
+    size_t maxEnumCount = 0;
+    for (auto elt : enumDecl->getAllElements()) {
+      if (!elt->getArgumentInterfaceType()) {
+        continue;
+      }
+      if (elt->isIndirect()) {
+        continue;
+      }
+      auto payloadTy = Ty.getEnumElementType(elt, Module);
+      fieldsCount = 0;
+      countNumberOfInnerFields(fieldsCount, Module, payloadTy);
+      if (fieldsCount > maxEnumCount) {
+        maxEnumCount = fieldsCount;
+      }
+    }
+    fieldsCount = fieldsCountBefore + maxEnumCount;
+  }
+}
+
+size_t TypeConverter::countNumberOfFields(SILType Ty) {
+  size_t fieldsCount = 0;
+  countNumberOfInnerFields(fieldsCount, M, Ty);
```

Comment:
I think countNumberOfInnerFields() should just return a count instead of modifying an out parameter by reference. Then you won't need a separate countNumberOfInnerFields() at all, TypeConverter::countNumberOfFields() could just be recursive.